+paradise/macros

xeno-by opened the pull request.

paulp committed:
Rewrote FastTrack for clarity.

We can say what we wish to say with more directness
and with fewer vars, levels of indirection, public members,
and implicit conversions.

xeno-by committed:
the scanner is now less eager about deprecations

When healing braces it isn't very useful to report deprecation warnings,
especially since this process is just simple context-free skimming, which
can't know about what positions can accept what identifiers.

xeno-by committed:
Moves annotationError outside typedAnnotation

This refactoring allows everyone from the compiler, e.g. the macro engine
which expands annotation-emitting macros, to produce annotation errors.

xeno-by committed:
fixes the typedIdent problem for good

Previous attachment retaining fix was only working for Idents which
get turned into Selects. Now it works for all transformations applied
to Idents (e.g. when an ident refers to something within a package obj).

retronym committed:
SI-6745 Fix <init> lookup

We should only consult the decls of the enclosing
class. Members of the self type, enclosing scopes,
or imports should not be considered.

xeno-by committed:
adds Tree.nonEmpty

So that trees become consistent with isEmpty on lists and options.

paulp committed:
SI-6846, regression in type constructor inference.

In 658ba1b4e6 some inference was gained and some was lost.
In this commit we regain what was lost and gain even more.
Dealiasing and widening should be fully handled now, as
illustrated by the test case.

paulp committed:
Fix and simplify typedTypeConstructor.

Investigating the useful output of devWarning (-Xdev people,
it's good for you) led back to this comment:

  "normalize to get rid of type aliases"

You may know that this is not all the normalizing does.
Normalizing also turns TypeRefs with unapplied arguments
(type constructors) into PolyTypes. That means that when
typedParentType would call typedTypeConstructor it would
find its parent had morphed into a PolyType. Not that it
noticed; it would blithely continue and unwittingly discard
the type arguments by way of appliedType (which smoothly
logged the incident, thank you appliedType.)

The simplification of typedTypeConstructor:

There was a whole complicated special treatment of AnyRef
here which appears to have become unnecessary. Removed special
treatment and lit a candle for regularity.

Updated lots of tests regarding newly not-so-special AnyRef.

paulp committed:
Removed dead implementation.

Another "attractive nuisance" burning off time until I
realized it was commented out.

paulp committed:
Cleaning up type alias usage.

I determined that many if not most of the calls to .normalize
have no intent beyond dealiasing the type. In light of this I
went call site to call site knocking on doors and asking why
exactly they were calling any of

  .normalize
  .widen.normalize
  .normalize.widen

and if I didn't like their answers they found themselves
introduced to 'dropAliasesAndSingleTypes', the recursive widener
and dealiaser which I concluded is necessary after all.

Discovered that the object called 'deAlias' actually depends
upon calling 'normalize', not 'dealias'. Decided this was
sufficient cause to rename it to 'normalizeAliases'.

Created dealiasWiden and dealiasWidenChain.

Dropped dropAliasesAndSingleTypes in favor of methods
on Type alongside dealias and widen (Type#dealiasWiden).
These should reduce the number of "hey, the type alias doesn't work" bugs.

paulp committed:
Shored up a hidden dealiasing dependency.

Like the comment says:

// This way typedNew always returns a dealiased type. This
// used to happen by accident for instantiations without type
// arguments due to ad hoc code in typedTypeConstructor, and
// annotations depended on it (to the extent that they worked,
// which they did not when given a parameterized type alias
// which dealiased to an annotation.) typedTypeConstructor
// dealiases nothing now, but it makes sense for a "new" to
// always be given a dealiased type.

PS:
Simply running the test suite is becoming more difficult all
the time. Running "ant test" includes time consuming activities
of niche interest such as all the osgi tests, but test.suite
manages to miss the continuations tests.

xeno-by committed:
renames c.fresh to c.freshName

xeno-by committed:
Backport from paradise/macros

Applies improvements to the macro engine from paradise/macros
to provide robustness and uniformity to different entrances of macro
expansion + to prepare for the unavoidable typemacro merge!

xeno-by committed:
enclosures are now strongly typed and are no longer vals

xeno-by committed:
adds c.introduceTopLevel

The first in the family of mutators for the global symbol table,
`introduceTopLevel` is capable of creating synthetic top-level
classes and modules.

The addition of nme.EMPTY_PACKAGE_NAME is necessary to let
programmers insert definitions into the empty package. That's explicitly
discouraged in the docs, but at times might come in handy.

This patch introduce workarounds to avoid incompatibilities with SBT.
First of all SBT doesn't like VirtualFiles having JFile set to null.
Secondly SBT gets confused when someone depends on synthetic files
added by c.introduceTopLevel.

Strictly speaking these problems require changes to SBT, and that will be
done later. However the main target of the patch is paradise/macros,
which needs to be useful immediately, therefore we apply workarounds.

xeno-by committed:
SI-6859 type macros no longer crash sbt

By mimicking alias types, macro types no longer trip a sanity check
in sbt 0.12 and 0.13: https://github.com/harrah/xsbt/blob/ef39aeb9c1ea443c17356695758a31152910c4a7/compile/interface/src/main/scala/xsbt/API.scala#L221.

This workaround makes me sad but it's the only option to make type macros
immediately useful. Principled fix pending.

xeno-by committed:
macro expansions are now auto-duplicated

The fix still requires macro developers to be careful about sharing trees
by references, because attributed DefTrees will still bring trouble.

However this is an improvement, because it doesn't make matters worse
and automatically fixes situations similar to one in the test.

A much more thorough discussion with a number of open questions left:
http://groups.google.com/group/scala-internals/browse_thread/thread/492560d941b315cc

xeno-by committed:
org.scala-lang.macro-paradise

xeno-by committed:
type macros beta v2

No spec yet, just docs at http://docs.scala-lang.org/overviews/macros/typemacros.html.
List of known issues: https://issues.scala-lang.org/secure/IssueNavigator.jspa?mode=hide&requestId=11912.

xeno-by commented:
Depends on a bunch of other pull requests opened against master. Please review those separately.

Currently only includes type macros. Quasiquotes will soon arrive.

review @paulp @adriaanm @retronym

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1249/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1967/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1249/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1967/

paulp committed:
Removed old pattern matcher.

paulp committed:
Merge pull request #1806 from paulp/pr/fasttrack

Rewrote FastTrack for clarity.

paulp committed:
Merge pull request #1820 from paulp/pr/what-old-patmat

Removed old pattern matcher.

paulp committed:
Merge pull request #1786 from retronym/ticket/6745-3

SI-6745 Fix <init> lookup

xeno-by commented:
removed https://github.com/scala/scala/pull/1810 from the dependencies. didn't touch anything else

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1256/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1974/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1256/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1974/

xeno-by committed:
changes the flags to not depend on partest

Due to some reason, partest always enables -deprecation. Since Paul has
just submitted a pull request, which removes this behavior, I'm updating
the flags to make sure this test works even after Paul's change.

paulp committed:
Made Symbol#associatedFile always return non-null

So we don't have to clutter everything with null checks.

paulp committed:
Merge pull request #1826 from adriaanm/paulp-pr-dealiasWiden

uniform dealiasing and widening

xeno-by committed:
Merge pull request #1811 from scalamacros/topic/tree-nonempty

adds Tree.nonEmpty

xeno-by committed:
Merge pull request #1815 from scalamacros/topic/fresh-name

renames c.fresh to c.freshName

xeno-by committed:
Merge pull request #1814 from scalamacros/topic/reflection

adds extractors for TypeName, TermName and Modifiers

xeno-by committed:
Merge pull request #1818 from scalamacros/topic/auto-duplicate-expansions

macro expansions are now auto-duplicated

xeno-by committed:
Merge pull request #1808 from scalamacros/topic/annotation-error

Moves annotationError outside typedAnnotation

paulp committed:
Merge pull request #1809 from scalamacros/topic/ident-attachments

fixes the typedIdent problem for good

xeno-by committed:
Merge remote-tracking branch 'origin/topic/enclosures' into paradise/macros

xeno-by committed:
Merge remote-tracking branch 'origin/topic/ident-deprecation-warnings' into paradise/macros

xeno-by committed:
Merge remote-tracking branch 'origin/topic/introduce-top-level' into paradise/macros

xeno-by committed:
Merge remote-tracking branch 'origin/topic/pre-typemacros' into paradise/macros

Conflicts:
	src/compiler/scala/reflect/macros/runtime/Enclosures.scala

xeno-by commented:
resubmitted in https://github.com/scala/scala/pull/1832
